v2/git-initializer/e2e/e2e.go

```diff
@@ -0,0 +1,182 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"time"
+
+	"github.com/ghodss/yaml"
+
+	"github.com/brigadecore/brigade/sdk/v2"
+	"github.com/brigadecore/brigade/sdk/v2/authx"
+	"github.com/brigadecore/brigade/sdk/v2/core"
+	"github.com/brigadecore/brigade/sdk/v2/restmachinery"
+	"github.com/pkg/errors"
+)
+
+type testcase struct {
+	name          string
+	projectFile   string
+	terminalPhase core.WorkerPhase
+}
+
+var testcases = []testcase{
+	{
+		name:          "GitHub - Public Repo - https",
+		projectFile:   "github-public-https.yaml",
+		terminalPhase: core.WorkerPhaseSucceeded,
+	},
+	// TODO: once we have mech for injecting private ssh key in secure manner
+	// {
+	// 	name:          "GitHub - Private Repo - ssh",
+	// 	projectFile:   "github-private-ssh.yaml",
+	// 	terminalPhase: core.WorkerPhaseSucceeded,
+	// },
+	{
+		name:          "GitHub - Public Repo - submodules",
+		projectFile:   "github-init-submodules.yaml",
+		terminalPhase: core.WorkerPhaseSucceeded,
+	},
+	{
+		name:          "GitHub - Malformed URL",
+		projectFile:   "github-malformed-url.yaml",
+		terminalPhase: core.WorkerPhaseFailed,
+	},
+}
+
+func main() {
+	for _, tc := range testcases {
+		err := runTest(tc)
+		if err != nil {
+			log.Fatalf("Test %q failed: %s", tc.name, err)
+		}
+	}
+}
+
+func runTest(tc testcase) error {
+	ctx, cancel :=
+		context.WithTimeout(context.Background(), 120*time.Second)
+	defer cancel()
+
+	// TODO: send in/parameterize this value
+	apiServerAddress := "https://localhost:7000"
+	// TODO: send in/parameterize this value
+	rootPassword := "F00Bar!!!"
+	apiClientOpts := &restmachinery.APIClientOptions{
+		AllowInsecureConnections: true,
+	}
+	filename := fmt.Sprintf("../testdata/%s", tc.projectFile)
+
+	log.Printf("Running test %q...\n", tc.name)
+
+	projectBytes, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return errors.Wrapf(err, "error reading project file %s", filename)
+	}
+
+	if projectBytes, err = yaml.YAMLToJSON(projectBytes); err != nil {
+		return errors.Wrapf(err, "error converting file %s to JSON", filename)
+	}
+
+	project := core.Project{}
+	if err = json.Unmarshal(projectBytes, &project); err != nil {
+		return errors.Wrapf(err, "error unmarshaling project file %s", filename)
+	}
+
+	authClient := authx.NewSessionsClient(
+		apiServerAddress,
+		"",
+		apiClientOpts,
+	)
+
+	token, err := authClient.CreateRootSession(ctx, rootPassword)
+	if err != nil {
+		return err
+	}
+	tokenStr := token.Value
+
+	client := sdk.NewAPIClient(
+		apiServerAddress,
+		tokenStr,
+		apiClientOpts,
+	)
```

Comment:
Excellent use of the SDK. This belongs in the textbook. 😉

Reply:
🍏 📖